Singular support for Verus

channy412 opened the pull request:
When a `proof` function is annoted with `verifier(integer_ring)`, Singular uses the Groebner basis for the ideal membership problem. This can be used only with `int` parameters. (Bounded integers are not supported now due to possible unsoundness).

For example, consider
```
    #[verifier(integer_ring)]
    #[proof]
    fn test2(a:int, s:int, R:int, M:int, RR:int, R_INV:int) {
        requires([
            (a * R - RR * s) % M == 0,
            (R_INV * R - 1) % M == 0,
            (RR - R * R % M) == 0,
        ]);
        ensures( (a - s*R) % M == 0);
    }
```
First, it registers all polynomials from requires clauses as a basis. These basis forms an ideal, and the Groebner basis is computed from it. If `ensures` polynomial is reduced to 0, it returns `Valid`, otherwise, it returns `invalid`.





This also includes some changes in CI. Test cases that use Singular are marked as `ignore`, but CI will run this every time.

(This PR is expected to be merged after https://github.com/secure-foundations/verus/pull/131)

yizhou7 commented on source/air/src/printer.rs:
```diff
@@ -495,3 +495,166 @@ impl NodeWriter {
+        ExprX::Const(Constant::Nat(n)) => n.to_string(),
+        ExprX::Var(x) => x.to_string(),
+        ExprX::Binary(BinaryOp::EuclideanMod, lhs, rhs) => {
+            // x % y ->  x - y*tmp
```
might want to use consistent temp variables for the same x and y

channy412 replied:
You're right, thank you!

yizhou7 commented on source/air/src/printer.rs:
```diff
@@ -495,3 +495,166 @@ impl NodeWriter {
+            format!("(({}))", ss.join(sop))
+        }
+        ExprX::Apply(fname, exprs) => {
+            if "Mul" == (**fname).as_str() {
```
maybe just do the translation in place, same applies to `EucMod`.

channy412 replied:
That's a good point, but after changing codes in that way, I just learned that constant multiplication still ends in that old path. So I think I will keep it this way.

yizhou7 commented on source/vir/src/well_formed.rs:
```diff
@@ -306,6 +318,70 @@ fn check_function(ctxt: &Ctxt, function: &Function) -> Result<(), VirErr> {
+        }
+        if let Some(body) = &function.x.body {
+            crate::ast_visitor::expr_visitor_check(body, &mut |expr| {
+                match &expr.x {
```
we might want to check the function calls in the pre/post conditions don't return bounded integers.

channy412 replied:
Thanks! I added more checks

utaal commented on source/air/src/ast.rs:
```diff
@@ -181,4 +181,5 @@ pub enum CommandX {
     SetOption(Ident, Ident), // set-option option value (no colon on the option)
     Global(Decl),            // global declarations
     CheckValid(Query),       // SMT check-sat (reporting validity rather than satisfiability)
+    SingularCheckValid(Vec<String>, Vec<Expr>, Vec<Expr>, Span), // check integer-ring proof using Singular: SingularCheckValid{params, ensures, requires, span}
```
I'd rather we didn't special-case on singular here.
* Why cannot this carry a `Query`? Can we define a `Query` type for singular, if there are differences?
* Depending on that we may want a separate variant (for solvers with a different kind of Query) or a `Solver` field on `CheckValid`.

channy412 replied:
I can make it fit into `Query` and make an additional boolean field(`is_singular`) in `QueryX`.

utaal commented on source/air/src/context.rs:
```diff
@@ -33,9 +35,11 @@ pub(crate) struct AxiomInfo {
 pub enum ValidityResult {
     Valid,
     Invalid(Model, Error),
+    SingularInvalid(Error),
```
Again, I don't like special-casing, as we'll need to handle them one by one.

Is there a reason why this cannot reuse `Invalid`? Maybe with an optional `Model` on `Invalid`?

channy412 replied:
Sure, I will try with `Option<Model>`

utaal commented on source/air/src/context.rs:
```diff
@@ -33,9 +35,11 @@ pub(crate) struct AxiomInfo {
     Canceled,
     TypeError(TypeError),
     UnexpectedSmtOutput(String),
+    UnexpectedSingularOutput(String),
```
Can we turn these both into `UnexpectedOutput`?

utaal commented on source/air/src/context.rs:
```diff
@@ -314,6 +318,47 @@ impl Context {
+        self.air_final_log.log_node(&nodes);
+
+        let query = singular_printer(vars, enss, reqs);
+        self.air_initial_log.comment(&query);
+        self.air_middle_log.comment(&query);
+        self.air_final_log.comment(&query);
```
I'm confused as what's happening here? Is this appending the singular query as a comment to the air file?
If so, why have the vars, requires, and ensures been printed as s-expr nodes? Is z3 supposed to see them?

channy412 replied:
I thought air log as an incorporated log file for all solvers, and imagined invoking Singular when `singular-check-valid` is encountered.  Would it be better to log the Singular part into a separate file?

utaal commented on source/air/src/main.rs:
```diff
@@ -126,6 +126,10 @@ pub fn main() {
             }
+            ValidityResult::SingularInvalid(..) => {
+                count_errors += 1;
+                println!("SingularInvalid in air main - Singular error at: TODO");
```
There's a leftover TODO here, which I imagine is turning this into proper error reporting instead of a println.

utaal commented on source/air/src/main.rs:
```diff
@@ -140,10 +144,16 @@ pub fn main() {
         if matches!(**command, CommandX::CheckValid(..)) {
             air_context.finish_query();
         }
+        if matches!(**command, CommandX::SingularCheckValid(..)) {
+            air_context.finish_query();
+        }
```
As discussed above, this is redundant with 151-153 due to having a separate CommandX variant.

utaal commented on source/air/src/printer.rs:
```diff
@@ -495,3 +495,182 @@ impl NodeWriter {
     NodeWriter::new().node_to_string_indent(&"".to_string(), node)
 }
+
+pub fn singular_query_to_node(vars: Vec<String>, enss: Vec<Expr>, reqs: Vec<Expr>) -> Node {
```
Is the output of this actually ever used by any of the solvers?

utaal commented on source/air/src/printer.rs:
```diff
@@ -495,3 +495,182 @@ impl NodeWriter {
+        ExprX::Multi(op, exprs) => {
+            let mut ss = vec![];
+            let sop = match op {
+                MultiOp::Add => ") + (",
```
Are these parenthesis starting/ending sub-expressions? Who's opening the first and closing the second?

utaal commented on source/air/src/printer.rs:
```diff
@@ -495,3 +495,182 @@ impl NodeWriter {
+                }
+            }
+        }
+        _ => panic!("unsupported singular expression: {:?}", expr),
```
Can we fail more gracefully here? Or at least print a span? That may be easier if this happens in `rust_verify`. (Unless this is a redundant check of one done in an upper layer.)

channy412 replied:
I guess I didn't use vir::def::MUL, because air does not depend on vir. (If I move these to `rust_verify` this would be solved as well)

utaal commented on source/docs/non_linear_arithmetic/non_linear_basic.rs:
```diff
@@ -0,0 +1,136 @@
+// testcases and examples are ported from below two files 
```
What's the purpose of this file? Is there a specific reason it isn't part of the tests?

channy412 replied:
I moved to docs because this lemma `LemmaMulUpperBound` failed on my machine, but it seems to work in the CI machine. (I will move to `examples`.)

Since `integer_ring` is available with `int`, users need to check bounds when they want to prove properties about bounded integers. That file is to show an example of that usage.

utaal commented on source/rust_verify/src/verifier.rs:
```diff
@@ -253,6 +253,7 @@ impl Verifier {
             (std::time::Duration::from_secs(2), report_fn)
         };
         let is_check_valid = matches!(**command, CommandX::CheckValid(_));
+        let is_singular = matches!(**command, CommandX::SingularCheckValid(..));
```
How are recommends checks handled in singular? Are they ignored?

channy412 replied:
I think the recommends check will simply run the same thing on Singular.

utaal commented on source/vir/src/sst_to_air.rs:
```diff
@@ -1621,26 +1622,60 @@ pub fn body_stm_to_air(
-        ]
+    if is_integer_ring {
+        // parameters, requires, ensures to Singular Query
+        let mut vars: Vec<String> = vec![];
```
Could these be AIR `Ident`s?

utaal commented on source/vir/src/well_formed.rs:
```diff
@@ -306,6 +318,102 @@ fn check_function(ctxt: &Ctxt, function: &Function) -> Result<(), VirErr> {
+        if function.x.ensure.len() != 1 {
+            return err_str(
+                &function.span,
+                "integer_ring mode: single ensure expression is assumed now",
```
"only a single ensures is allowed in integer_ring mode"

utaal commented on source/vir/src/well_formed.rs:
```diff
@@ -306,6 +318,102 @@ fn check_function(ctxt: &Ctxt, function: &Function) -> Result<(), VirErr> {
+                        _ => {
+                            return err_str(
+                                &function.span,
+                                "integer_ring mode ensure expression: `X % m == 0` or `X == Y` ",
```
Can you expand on this error message a bit?

utaal commented on source/vir/src/well_formed.rs:
```diff
@@ -306,6 +318,102 @@ fn check_function(ctxt: &Ctxt, function: &Function) -> Result<(), VirErr> {
+            } else {
+                return err_str(
+                    &function.span,
+                    "integer_ring mode ensures expression: operator `==` is assumed",
```
What do you mean that's assumed? That it is required?

channy412 replied:
I wanted to say that the most outer operator for `ensures` expression should be the equality operator  (not any functions that return boolean, especially shouldn't be inequalities)

utaal reviewed (changes requested):
I think you should squash all the changes now and force-push to this branch, as it will be hard to keep track of review items otherwise.

I may have more comments on the Singular emitter as I take a more in-depth look.

utaal commented:
Considering the code is mostly separate from the main lowering logic, the easiest way to make singular an optional dependency may be to use a rust "feature" to gate the code and the tests: https://doc.rust-lang.org/cargo/reference/features.html

channy412 commented:
@utaal Thank you very much for the detailed review! I will try to resolve them one by one.

channy412 commented on source/rust_verify/example/integer_ring.rs:
```diff
@@ -0,0 +1,125 @@
+// most of the codes are copied from: https://github.com/utaal/verified-nrkernel/blob/main/page-table/lib.rs
```
Thanks, will do!

channy412 commented on source/air/src/printer.rs:
```diff
@@ -495,3 +495,182 @@ impl NodeWriter {
+        vars2.push(v);
+    }
+
+    // gather polynomials that will be the basis of ideal
+    // for requires,  equality -> register ideal (rhs - lhs)
```
I agree that those parts will be better at `rust_verify`.   Thanks again for the reviews!

channy412 commented:
I refactored many parts of this singular support. 

First, I expanded `CommandsWithContext`'s `is_spinoff` to be an enum, to indicate whether it is for default z3, spinoff z3, or Singular. I also merged singular commands with SMT commands.

I moved out the main logic part into `rust_verify`, and subsequently made it into a `singular` feature for conditional compilation. 

Instead of using the AIR log file, I moved out the logging part into a `.singular` log file, which can be directly run with Singular. (Similar to running `root.smt2` with z3 directly)

Then I removed most of the `panic` during the translation logic into proper error reporting.

utaal reviewed (approved):
This is great. Thanks a lot for the changes!
